ui/insight/src/ctrl.ts

```diff
@@ -1,173 +1,160 @@
+      [],
+      env.ui.dimensionCategs.map(c => c.items)
+    );
+    this.metrics = Array.prototype.concat.apply(
+      [],
+      env.ui.metricCategs.map(c => c.items)
+    );
+
+    this.vm = {
+      metric: this.findMetric(this.env.initialQuestion.metric)!,
+      dimension: this.findDimension(this.env.initialQuestion.dimension)!,
+      filters: this.env.initialQuestion.filters,
+      loading: true,
+      broken: false,
+      answer: null,
+      panel: Object.keys(env.initialQuestion.filters).length ? 'filter' : 'preset',
+    };
+  }
+
+  private findMetric = (key: string) => this.metrics.find(x => x.key === key);
+
+  private findDimension = (key: string) => this.dimensions.find(x => x.key === key);
+
+  setPanel(p: 'filter' | 'preset') {
     this.vm.panel = p;
-    m.redraw();
-  }.bind(this);
+    this.redraw();
+  }
 
-  var reset = function () {
+  reset() {
     this.vm.metric = this.metrics[0];
     this.vm.dimension = this.dimensions[0];
     this.vm.filters = {};
-  }.bind(this);
-
-  var askQuestion = throttle(
-    1000,
-    function () {
-      if (!this.validCombinationCurrent()) reset();
-      this.pushState();
-      this.vm.loading = true;
-      this.vm.broken = false;
-      m.redraw();
-      setTimeout(
-        function () {
-          m.request({
-            method: 'post',
-            url: env.postUrl,
-            data: {
-              metric: this.vm.metric.key,
-              dimension: this.vm.dimension.key,
-              filters: this.vm.filters,
-            },
-            deserialize: function (d) {
-              try {
-                return JSON.parse(d);
-              } catch (e) {
-                throw new Error(d);
-              }
-            },
-          }).then(
-            function (answer) {
-              this.vm.answer = answer;
-              this.vm.loading = false;
-            }.bind(this),
-            function () {
-              this.vm.loading = false;
-              this.vm.broken = true;
-              m.redraw();
-            }.bind(this)
-          );
-        }.bind(this),
-        1
-      );
-    }.bind(this)
-  );
-
-  this.makeUrl = function (dKey, mKey, filters) {
-    var url = [env.pageUrl, mKey, dKey].join('/');
-    var filters = Object.keys(filters)
+  }
+
+  askQuestion = throttle(1000, () => {
+    if (!this.validCombinationCurrent()) this.reset();
+    this.pushState();
+    this.vm.loading = true;
+    this.vm.broken = false;
+    this.redraw();
+    setTimeout(() => {
+      xhr
+        .json(this.env.postUrl, {
+          method: 'post',
+          headers: { 'Content-Type': 'application/json' },
+          body: JSON.stringify({
+            metric: this.vm.metric.key,
+            dimension: this.vm.dimension.key,
+            filters: this.vm.filters,
+          }),
+        })
+        .then(
+          (answer: Chart) => {
+            this.vm.answer = answer;
+            this.vm.loading = false;
+            this.redraw();
```

Comment:
The original code doesn't seem to call `redraw` after receiving a successful response. Maybe it gets called somewhere in mithril internals?
I added a `redraw` call because the ui clearly needs to be redrawn once data is received.

Reply:
yes mithril has magic redraw after XHR processing, snabbdom doesn't.